vpp.man
.TH VPP 1
.SH NAME
vpp \- Verilog Pre Processor
.SH SYNOPSIS
.I vpp
file1 [file2 ...]
.SH OPTIONS
.IP
none currently defined
.SH DESCRIPTION
.I vpp
provides an improved set of pre-processor features not provided by
the original Verilog language.  New features include `ifndef,
`if <expression>, `switch, `for, `let <var>=<expression>, and 
`while.  The post operators ++ and -- are supported, not only in the
epressions of the extensions, but also in the Verilog code itself.

Vpp parses `defines and these values can be used by other preprocessor
commands.  Vpp traverses `include hierarchies, and examines files in the
order specified.

.SH Extension `ifdef and `ifndef
`ifdef functions the same as the original `ifdef preprocessor command.  Additionally,
`ifdef and `ifndef leave the preprocessor commands embedded in the source code,
unmodified.  Future versions of vpp will allow you to specify if you want
the preprocessor commands stripped for `ifdef and `ifndef.  Note that `ifndef
is converted to a `ifdef, by simply reversing the fields between `ifndef and `else
with the fields between `else and `endif.
.SH Extension `if
The `if expression allows expressions to be evaluated, and if true, will active the
region of code between the `if and the matching `endif (or the matching `else, if it
exists).  See the "expressions" section below for an explanation of what expressions
are legitimate in vpp.
.SH Extension `switch
The `switch extension, and its matching `endswitch are used to conditionally include
code with vpp.  The `case identifies a region that is evaluated against the `switch
expression to see if the `case's region is active. Example :
.nf
`let x=5
`switch (x)
`case 1
   if x is one, print out this line
`case 5
   if x is five, print out this line
`default
   if x does not match any above cases, print out this line
`endswitch
.fi

The above example prints out the `case 5 region (it is the region from the current `case
to the next `case, `default, or `endswitch that matches this `case (as nesting of `switches
is allowed).
.SH Extension `for
The `for extension allows the emulation of the "generate" statement of VHDL.  The syntax
of the `for statement is : `for (identifier=expression ; expression ; identifier=expression).
The code between the `for and its matching `endfor is the region that is repeated until
the middle expression is satisfied.  Nesting of expression is allowed.  Below is an example
of a `for loop

.nf
`for (i=0; i<4; i++)
   i=`i
`endfor
.fi

.SH Extension `while
The `while extension has the following syntax : `while expression.
The code between the `while and its matching `endwhile is the region that is repeated until
the expression is not true. `while's can be nested.
.SH Extension `let
The `let is used to assign variables in vpp. The syntax for the `let is
`let identifier = expression.  Variables can be integers, reals, or strings.
The type of the variable is determined by the resulting operations.  Any
identifier that is assigned with a `let is available for evaluation in
any subsequent expressions.  Additionally, the ascii equivalent of a variable
is available in the text simply by prepending the identifier by a backtick (`).
Anywhere this might cause confusion, a new operator called append (::) is
now available.  E.g. a::`i::b results in a5b, if i was equal to 5.
 The following are
legitimate `let assignments.  See "expressions" for more information.

.nf
`let x=5
`let x=5;
`let x=1.2
`let x="abcde"
`let r=10
`let circumference=355/113 * 2 * r
`let pi=355/113
`let area = pi * r**2
.fi

Currently, operators on strings are not defined (i.e. you cannot add two strings together).
.SH expressions
Expression can be simple or complex, and can include the following operators :
numeric, logical bitwise, equality and functions.  Precedence rules follow that of Verilog
(e.g. multiplication has higher precedence than addition).  The use of brackets
("(" and ")") allows user specification on order of evaluation.

.SH numeric operators
Numeric operators include : addition (+), subtraction (-), multiplication (*),
division (/), modulus (%), power (**).
.SH logical operators
Logial operators include : logical AND (&&), logical OR (||), logical NOT (!)
.SH bit operators
Bit operators include : bitwise XOR (^), bitwise AND (&), bitwise OR (|), shift right (>>),
shift left (<<)
.SH equality operators
Equality operators include : equality (==), inequality (!=), greater than (>), less than (<),
greater than or equal to (>=), less than or equal to (<=)
.SH function
The functions include : log 2 (LOG2()), round up (CEIL()), round down (FLOOR()), round
to nearest value (ROUND()),
max of two numbers (MAX()), min of two numbers (MIN()), odd (ODD()), even (EVEN()),
absolute value (ABS()).
.SH string expansion
Any variable that is created (or defined with the `define), can be expanded as a string
in the Verilog code.  This is accomplished by prepending the backtick (`) before the
variable you wish to expand.  Anywhere this can cause confusion, you can use the
append operator (::).  Examples of string expansions include :

.nf
`let x=134*531
an expansion of a`x
an embedded expansion a`x::b
.fi

This would result in the following code :

.nf
an expansion of a71154
an embedded expansion a71154b
.fi


.SH USAGE EXAMPLES
See the directory src/EXAMPLES that has been provided with the source for vpp.

.SH BUGS

.SH "AUTHOR"
.nf
Original program written by
    Himanshu M. Thaker (hmthaker@tddcae99.fnts.com)
.fi
.SH FURTHER INFORMATION


CHANGELOG
2002-02-02  Jimen Ching  <jching@flex.com>

	* Makefile.in:
	Generate tarball with directory of vbpp-VERNAME.

	* README: Update for new release.

2002-01-31  Jimen Ching  <jching@flex.com>

	* vpp.l:
	Detect a directive specifier ` in a name token is an error.

2002-01-30  Jimen Ching  <jching@flex.com>

	* EXAMPLES/1exout-template, EXAMPLES/x_arguments.vpp, EXAMPLES/x_macro.vpp, Makefile.in:
	New test samples for new macro function feature.

	* vpp.l: Support macro functions.

2002-01-28  Jimen Ching  <jching@flex.com>

	* vpp.l:
	Handle long lines more gracefully, do write beyond end of buffer.

	* common.h:
	Increase the maximum line length supported.

	* vpp.y:
	Stop parsing after 100 parse errors.

	* vpp.c:
	Add -h option to print out command line options list.

	* Makefile.in:
	Generate yacc parser before lexor, since the lexor needs some yacc
	symbols.

	* create.c, vpp.l:
	Store macro definition only when it is active.

2002-01-27  Jimen Ching  <jching@flex.com>

	* vpp.y:
	Do not output comment starters.  That is done in lexor.

2002-01-26  Jimen Ching  <jching@flex.com>

	* vpp.l: Fix typo in vectornets.
	Expand `name macros when C style preprocessing is enabled.
	Store comments in MPP state, otherwise, it is outputed too early.

	* vpp.c:
	Add option to silence status message.

	* create.c:
	Do not evaluate expression for `ifdef/`ifndef directives.

2002-01-20  Jimen Ching  <jching@flex.com>

	* create.c, vpp.c, yacc_stuff.h:
	Add command line option to output `line directive and process `ifdef
	conditionals like C preprocessors.

	* vpp.l:
	nl_count was incremented too soon.

	* configure, configure.in:
	Make output of line directive a command line option instead of
	configure option.

	* EXAMPLES/1exout-template, EXAMPLES/2exout-template, vpp.l:
	Code cleanup.
	Made error/warning messages not as verbose.

2002-01-18  Jimen Ching  <jching@flex.com>

	* vpp.l:
	Allow `keyword for INITIAL and COPY states only.  Addition of comment
	states caused processing of these directives if not restricted.

	* common.h, proto.h, vpp.c, vpp.l:
	Add support for +incdir+ and -D options.

	* EXAMPLES/1exout-template, EXAMPLES/x_cmddefine.vpp, EXAMPLES/x_hdr.vhh, EXAMPLES/x_inc.vpp, Makefile.in:
	New tests for +incdir+ and -D options.

2002-01-13  Jimen Ching  <jching@flex.com>

	* vpp.l:
	Use the lex facilities to handle C style comments.
	Other code cleanup.

	* EXAMPLES/0exout-template, EXAMPLES/1exout-template, EXAMPLES/2exout-template, vpp.c, vpp.l, yacc_stuff.h:
	Support generating `line directive.
	Fixed bug where `include directive incorrectly tests the first include.
	Added `line directive to the beginning of parsing.
	Increased limit for include depth and `line output.

	* Makefile.in:
	Do not use symbolic link to run through examples.

	* configure, configure.in:
	Support generating `line directives.

2001-03-11  Jimen Ching  <jching@flex.com>

	* README:
	Updated for release 1.0.4.

	* CHANGELOG:
	Updated new checkin comments.

--------------------------------------------------------------------------------

2001-01-25  Jimen Ching  <jching@flex.com>

	* Makefile.in, vpp.c, vpp.y, yacc_stuff.h:
	Return non-zero when parsing error detected.

2000-11-08  Jimen Ching  <jching@flex.com>

	* Makefile.in:
	Delete EXAMPLES only if it is a symbolic link.

2000-08-20  Jimen Ching  <jching@flex.com>

	* common.h:
	More ANSI conformance.

2000-08-11  Jimen Ching  <jching@flex.com>

	* create.c:
	Handle qouted string in expression evaluation.

2000-07-29  Jimen Ching  <jching@flex.com>

	* EXAMPLES/1exout-template, EXAMPLES/x_define.vpp, Makefile.in:
	New example code to test preprocessor macro define handling.

	* create.c:
	Use new get_define_value for error detection.
	Skip 'for' loop if expression is false on entry.

	* proto.h, vpp.l:
	Changed get_define_value API to allow error detection.

2000-07-09  Jimen Ching  <jching@flex.com>

	* EXAMPLES/simple_switch.vpp, EXAMPLES/x_swfw.vpp, create.c:
	The outer active state was not taken into account when processing `endif.

--------------------------------------------------------------------------------

Thu Dec 30 21:06:21 1999 Emile van Rooyen <emile.van.rooyen@microchip.com>

	* [vpp.y]
	Prevent stack overflows.

Thu Jul  1 21:52:30 1999  Jimen Ching  <jching@flex.com>

        * [create.c]
	Fixed bug in eval_stack where `endfor is not skipped when it should be.

	* [configure.in Makefile.in]
	Code cleanup.

	* [EXAMPLES]
	Added more examples for testing.

Fri Feb 13 21:08:00 1998  Jimen Ching  <jching@flex.com>

        * [create.c]
	Added UPIPE, UNOR, UXNOR.

Mon Feb  9 21:43:43 1998  Jeffrey Ebert  <jebert@hns.com>

	* [proto.h vpp.l]
	Changed call from rint to round_int.

Sat Jan 10 11:08:39 1998  Jimen Ching  <jching@flex.com>

	* [create.c]
	Added UAND, UNAND, UNOR.


vpp.c
/* Copyright (C) 1996 Himanshu M. Thaker

This file is part of vpp.

Vpp is distributed in the hope that it will be useful,
but without any warranty.  No author or distributor
accepts responsibility to anyone for the consequences of using it
or for whether it serves any particular purpose or works at all,
unless he says so in writing.

Everyone is granted permission to copy, modify and redistribute
vpp, but only under the conditions described in the
document "vpp copying permission notice".   An exact copy
of the document is supposed to have been given to you along with
vpp so that you can know how you may redistribute it all.
It should be in a file named COPYING.  Among other things, the
copyright notice and this notice must be preserved on all copies.  */

/*
 * Program : vpp
 * Author : Himanshu M. Thaker
 * Date : Apr. 16, 1995
 * Description :
 */

#include "common.h"
#include "yacc_stuff.h"
#include "proto.h"
#include <string.h>
#include <stdio.h>

struct vpp_filename
	{
	struct vpp_filename *next;   
	char *name;
	};
struct vpp_filename *vpp_incdir;
int output_quite;

int
main(int argc, char *argv[])
	{
	int i, arglen;
	struct vpp_filename *t, *filehead, *filecurr, *idcurr;

#ifdef YYDEBUG
	yydebug = TRUE;
#endif

	/* Default to old behavior. */
	output_line_directive = FALSE;
	output_ifdef_directive = TRUE;

	filehead = filecurr = idcurr = NULL;
	for (i = 1; i < argc; ++i)
		{
		arglen = strlen(argv[i]);
		if (arglen >= 8 && !strncmp(argv[i],"+incdir+",8))
			{
			if (arglen == 8)
				{
				/* Perhaps we don't even need to notify user. */
				fprintf(stderr, "** No path for +incdir+ in argument %d.\n", i);
				}
			else
				{   
				char *lname, *tok;

				lname = (char *)NMALLOC(arglen-8+1, char);
				strcpy(lname, argv[i]+8);
				tok = strtok(lname, "+");
				while (tok)
					{
					int toklen = strlen(tok);
					t = CALLOC(1, struct vpp_filename);
					t->name = NMALLOC(toklen+1, char);
					strcpy(t->name, tok);
					if (!vpp_incdir)
						vpp_incdir = t;
					else
						idcurr->next = t;
					idcurr = t;
					tok = strtok(NULL, "+");
					}
				free(lname);
				}
			}
		else if (arglen >= 2 && !strncmp(argv[i], "-D", 2))
			{
			if (arglen == 2)
				{
				/* Perhaps we don't even need to notify user. */
				fprintf(stderr, "** No define for -D in argument %d.\n", i);
				}
			else
				{
				char *dname, *pnt;

				dname = NMALLOC(arglen+8, char);
				sprintf(dname, "`define %s", argv[i]+2);
				pnt = dname+8;
				while (*pnt)
					{
					if (*pnt == '=')
						{
						*pnt = ' ';
						break;
						}
					pnt++;
					}
				store_define(dname);
				free(dname);
				}
			}
		else if (arglen >= 2 && !strncmp(argv[i], "-E", 2))
			{
			/* Enable C style preprocessing of ifdef and friends. */
			output_ifdef_directive = FALSE;
			}
		else if (arglen >= 2 && !strncmp(argv[i], "-L", 2))
			{
			/* Write out `line directives for debugging. */
			output_line_directive = TRUE;
			}
		else if (arglen >= 2 && !strncmp(argv[i], "-h", 2))
			{
			printf("syntax: %s [options] filename\n", argv[0]);
			printf("options:\n");
			printf("  +incdir+...+ Search directory for `include files.\n");
			printf("  -D           Define macro.\n");
			printf("  -E           Perform C style preprocessing.\n");
			printf("  -L           Output `line directive.\n");
			printf("  -h           Print help message and exit.\n");
			printf("  -q           Suppress status message.\n");
			return 0;
			}
		else if (arglen >= 2 && !strncmp(argv[i], "-q", 2))
			{
			/* Quite, take out unnecessary output. */
			output_quite = TRUE;
			}
		else /* filename only.. */
			{
			t = CALLOC(1,struct vpp_filename);
			t->name = NMALLOC(strlen(argv[i])+1, char);
			strcpy(t->name, argv[i]);
			if (!filehead)
				filehead = t;
			else
				filecurr->next = t;
			filecurr = t;
			}
        }

	/*
	 * check to see if the file can be opened for reading...
	 */
	if (!filehead)
		{
		yyin = stdin;
		current_file = NMALLOC(strlen("<stdin>")+1, char);
		strcpy(current_file, "<stdin>");

		nl_count = 1;
		if (!output_quite)
			fprintf(stderr, "Preprocessing '%s'\n", current_file);
		do_comment_count(TRUE, nl_count);
		yyparse();
		fclose(yyin);
		free(current_file);
		}
	else
		{
		filecurr = filehead;
		while (filecurr)
			{
			yyin = fopen(filecurr->name, "r");
      		if (yyin == NULL)
				{
        		fprintf(stderr, "** %s: open failed\n", filecurr->name);
				++yyerror_count;
				}
			else
				{
      			current_file = NMALLOC(strlen(filecurr->name)+1, char);
      			strcpy(current_file, filecurr->name);

				nl_count = 1;
				if (!output_quite) {
					fprintf(stderr, "vpp 1.2.0, see 'https://github.com/balanx/vbpp'\n");
					fprintf(stderr, "Preprocessing '%s'\n", current_file);
				}
				do_comment_count(TRUE, nl_count);
				yyparse();
				fclose(yyin);
				free(current_file);
				}

			t = filecurr;
			filecurr = filecurr->next;
			free(t->name);
			free(t);
			}

		filehead = NULL;
		}

	return yyerror_count;
	}

FILE *
incdir_fopen(char *name)
	{
	int len;
	struct vpp_filename *l;  
	char *path;
	FILE *fp = NULL;

	len = strlen(name);
	l = vpp_incdir;
	while (l)
        {
        path = (char *)malloc(strlen(l->name) + 1 + len + 1);
        sprintf(path, "%s/%s", l->name, name);
        fp = fopen(path, "r");
        free(path);
        if (fp)
			break;
		l = l->next;
        }

	if (!fp)
		fp = fopen(name, "r");
	return(fp);
	}
